# Find out hosting countries and providers (ISP)

- IP can be found with `socket`
- hostings are easy to find (`ipinfo` library, `ip2location` opensource library, hostings are already discovered by `whois`)
- ISP can be found with the help of ip2location https://www.ip2location.com/developers/python


In [2]:
import pandas as pd
import requests, json, os
from json.decoder import JSONDecodeError
import json

In [ ]:
data_url = "https://raw.githubusercontent.com/Teplitsa/CSRLab/main/data/2022_lab_index_report.csv"
index_report_data = pd.read_csv(data_url, encoding='utf-8')
data_url = "https://raw.githubusercontent.com/Teplitsa/CSRLab/main/data/2022_lab_selected_ngos_data.csv"
selected_ngos_data = pd.read_csv(data_url, encoding='utf-8')
concat_ngos = pd.merge(
    index_report_data,
    selected_ngos_data,
    how="outer",
    on=["ogrn", "website"],
    sort=False,
    left_index=False,
    right_index=False,
)

In [ ]:
concat_ngos.head()

ogrn         website  bin_robots  bin_sitemap  bin_fb  bin_vk  \
0  1026800000666    profobr68.ru         1.0          1.0     0.0     0.0   
1  1022900002861  проф-севмаш.рф         1.0          0.0     0.0     1.0   
2  1025200008140      vdpor52.ru         1.0          1.0     0.0     0.0   
3  1037739471220     orthonet.ru         0.0          0.0     0.0     0.0   
4  1085200004450   vdpo-sarov.ru         1.0          1.0     0.0     1.0   

   bin_ig  bin_ok  bin_youtube  bin_tiktok  ...  bin_socnet  \
0     0.0     0.0          0.0         0.0  ...    0.000000   
1     0.0     0.0          0.0         0.0  ...    0.166667   
2     0.0     0.0          0.0         0.0  ...    0.000000   
3     0.0     0.0          0.0         0.0  ...    0.000000   
4     0.0     0.0          0.0         0.0  ...    0.166667   

              regionName               minjustForm  \
0     Тамбовская область     Профессиональный союз   
1  Архангельская область     Профессиональный союз   
2  Нижегородская область  Общественная организация   
3                 Москва  Общественная организация   
4  Нижегородская область  Общественная организация   

                                           shortName     regDate  \
0  ТАМБОВСКАЯ ОБЛАСТНАЯ ОРГАНИЗАЦИЯ ОБЩЕРОССИЙСКО...  1938-11-10   
1                  МОО - ППО АО "ПО "СЕВМАШ" СУДПРОФ  1942-03-15   
2  НИЖЕГОРОДСКОЕ ОБЛАСТНОЕ ОТДЕЛЕНИЕ ВДПО, ВДПО Н...  1958-06-20   
3                              СОЮЗ ПИСАТЕЛЕЙ РОССИИ  1958-12-07   
4                                   САРОВСКОЕ "ВДПО"  1962-03-30   

                                           mainOkved  \
0               Деятельность профессиональных союзов   
1               Деятельность профессиональных союзов   
2  Деятельность прочих общественных организаций, ...   
3  Деятельность профессиональных членских организ...   
4  Деятельность прочих общественных организаций, ...   

                opfCombined  \
0   Профсоюзные организации   
1   Профсоюзные организации   
2  Общественные организации   
3  Общественные организации   
4  Общественные организации   

                                       okvedCombined  \
0               Деятельность профессиональных союзов   
1               Деятельность профессиональных союзов   
2  Деятельность прочих общественных организаций, ...   
3  Деятельность профессиональных членских организ...   
4  Деятельность прочих общественных организаций, ...   

                  vedSimple  websiteCreationDate  
0                 Профсоюзы  2016-01-15 19:31:34  
1                 Профсоюзы  2018-08-27 11:35:06  
2  Общественные организации  2016-09-28 12:05:06  
3                 Профсоюзы  2005-11-24 21:00:00  
4  Общественные организации   2021-04-15 8:10:11  

[5 rows x 27 columns]

I also need whois data from Gryadka:

In [ ]:
data_url = "https://raw.githubusercontent.com/Teplitsa/CSRLab/main/data/2021_nov_whois_check.csv"
whois_data = pd.read_csv(data_url, encoding='utf-8')
whois_data.tail()

domain_name     registrar        creation_date  \
13377          PORA-DOBRA.RU  RU-CENTER-RU  2018-08-02 15:53:23   
13378         EDUC-CENTER.RU  RU-CENTER-RU  2020-02-04 06:58:33   
13379             FONDNKN.RU  RU-CENTER-RU  2018-07-17 15:00:41   
13380  XN--J1AJCABV.XN--P1AI   RUCENTER-RF  2019-02-10 16:21:35   
13381          UNINOVAMED.RU      REGRU-RU  2018-10-10 09:23:27   

           expiration_date                             status  \
13377  2022-08-02 15:53:23  REGISTERED, DELEGATED, UNVERIFIED   
13378  2023-02-04 06:58:33  REGISTERED, DELEGATED, UNVERIFIED   
13379  2022-07-17 15:00:41    REGISTERED, DELEGATED, VERIFIED   
13380  2023-02-10 16:21:36  REGISTERED, DELEGATED, UNVERIFIED   
13381  2022-10-10 09:23:27  REGISTERED, DELEGATED, UNVERIFIED   

                                                     org         website  
13377                 Charity Foundation " TIME OF GOOD"   pora-dobra.ru  
13378  Avtonomnaya nekommercheskaya organizatsiya "Ts...  educ-center.ru  
13379                    Fund National Cultural Heritage      fondnkn.ru  
13380  Nekommercheskaya organizatsiya Respubliki Komi...       рфрпрк.рф  
13381                               ANO "MICLT "Uninova"   uninovamed.ru

In [ ]:
len(whois_data)

13382

In order to get hostings and providers we need to get IP addresses first:

In [ ]:
import socket

IP_addres = socket.gethostbyname('sovetveteranov01.ru')
print("IP Address is: " + IP_addres)

IP Address is: 5.23.51.195


`http://ip-api.com/json/[ip address]`

 45 requests per minute from an IP address.

SSL (HTTPS)

256-bit SSL encryption is not available for this free API. Please see our pro service.

In [ ]:
concat_ngos[concat_ngos['bin_ssl'] == 1.0]

ogrn              website  bin_robots  bin_sitemap  bin_fb  \
5     1024700001567         schoolvoa.ru         1.0          1.0     0.0   
9     1027800013790     открытыйгород.рф         1.0          1.0     1.0   
15    1020100003461  sovetveteranov01.ru         1.0          1.0     0.0   
16    1022200528470    altaidetfond22.ru         0.0          1.0     1.0   
20    1021600007582        profchelny.ru         1.0          0.0     0.0   
...             ...                  ...         ...          ...     ...   
9535  1211900002621             frprh.ru         1.0          1.0     1.0   
9537  1214700012845        anoecograd.ru         1.0          1.0     0.0   
9539  1213500013407           мызавас.рф         1.0          1.0     0.0   
9542  1217700396330           gorneks.ru         1.0          1.0     0.0   
9544  1212300052249      blagfond-zzh.ru         0.0          1.0     0.0   

      bin_vk  bin_ig  bin_ok  bin_youtube  bin_tiktok  ...  bin_socnet  \
5        0.0     0.0     0.0          0.0         0.0  ...    0.000000   
9        1.0     1.0     0.0          1.0         0.0  ...    0.666667   
15       0.0     1.0     0.0          0.0         0.0  ...    0.166667   
16       0.0     1.0     0.0          0.0         0.0  ...    0.333333   
20       0.0     1.0     0.0          1.0         0.0  ...    0.333333   
...      ...     ...     ...          ...         ...  ...         ...   
9535     0.0     0.0     0.0          0.0         0.0  ...    0.166667   
9537     0.0     0.0     0.0          0.0         0.0  ...    0.000000   
9539     0.0     0.0     0.0          0.0         0.0  ...    0.000000   
9542     0.0     0.0     0.0          0.0         0.0  ...    0.000000   
9544     0.0     0.0     0.0          0.0         0.0  ...    0.000000   

                 regionName               minjustForm  \
5     Ленинградская область  Общественная организация   
9           Санкт-Петербург                       NaN   
15        Республика Адыгея  Общественная организация   
16           Алтайский край         Общественный фонд   
20     Республика Татарстан     Профессиональный союз   
...                     ...                       ...   
9535     Республика Хакасия       Некоммерческий фонд   
9537  Ленинградская область                       NaN   
9539    Вологодская область       Некоммерческий фонд   
9542                 Москва                       NaN   
9544     Краснодарский край                       NaN   

                                              shortName     regDate  \
5                      ГАТЧИНСКАЯ ГОРРАЙОРГАНИЗАЦИЯ ВОА  1973-05-08   
9                                    СПБ ГО ВОО ВООПИИК  1976-05-15   
15    ОО ВЕТЕРАНОВ (ПЕНСИОНЕРОВ) ВОЙНЫ, ТРУДА, ВООРУ...  1987-03-06   
16    АЛТАЙСКОЕ КРАЕВОЕ ОТДЕЛЕНИЕ РОССИЙСКОГО ДЕТСКО...  1988-02-26   
20    НАБЕРЕЖНО-ЧЕЛНИНСКАЯ ТЕРРИТОРИАЛЬНАЯ ОРГАНИЗАЦ...  1989-09-09   
...                                                 ...         ...   
9535                                             ФРП РХ  2021-08-05   
9537                                      АНО "ЭКОГРАД"  2021-08-10   
9539                                     БФ "МЫ ЗА ВАС"  2021-08-13   
9542                                      АНО "ГОРНЭКС"  2021-08-24   
9544             БЛАГОТВОРИТЕЛЬНЫЙ ФОНД "ЗДРАВИЯ ЖЕЛАЮ"  2021-09-08   

                                              mainOkved  \
5     Деятельность прочих общественных организаций, ...   
9     Деятельность по охране исторических мест и зда...   
15    Деятельность прочих общественных организаций, ...   
16    Деятельность прочих общественных организаций, ...   
20                 Деятельность профессиональных союзов   
...                                                 ...   
9535  Деятельность по предоставлению займов промышле...   
9537  Научные исследования и разработки в области ес...   
9539  Предоставление прочих финансовых услуг, кроме ...   
9542                   Судебно-экспертная деятельность}   
9544  Предоста

In [ ]:
!curl http://ip-api.com/json/5.23.51.195

{"status":"success","country":"Russia","countryCode":"RU","region":"SPE","regionName":"St.-Petersburg","city":"St Petersburg","zip":"190005","lat":59.8889,"lon":30.3296,"timezone":"Europe/Moscow","isp":"TIMEWEB Co Ltd.","org":"TIMEWEB","as":"AS9123 TimeWeb Ltd.","query":"5.23.51.195"}

### How to get all the info:

* **hosting** domain name from already available `whois` dataset
* hosting **country** `IP2Location` or `ipinfo`. It is not clear which database has more information. I should use `IP2Location` as the first step since it is unlimited, and if it lacks information turn to `ipinfo`.
* **isp** - available only for websites with ssl in `ip-api`

*ipinfo* library for hostings and providers has a limit of 50,000 requests per month:

In [ ]:
!pip install ipinfo

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 1.1 MB 8.1 MB/s 
     |████████████████████████████████| 271 kB 66.0 MB/s 
     |████████████████████████████████| 94 kB 3.6 MB/s 
     |████████████████████████████████| 144 kB 55.8 MB/s 


IP2Location:

https://www.ip2location.com/database/db2-ip-country-isp

In [ ]:
!unzip IP2Location-Python-master.zip

Archive:  IP2Location-Python-master.zip
64295dd1daaa9145ebde966c57b995b3ba00c988
   creating: IP2Location-Python-master/
 extracting: IP2Location-Python-master/AUTHORS  
  inflating: IP2Location-Python-master/ChangeLog  
  inflating: IP2Location-Python-master/IP2Location.py  
  inflating: IP2Location-Python-master/LICENSE.TXT  
  inflating: IP2Location-Python-master/PKG-INFO  
  inflating: IP2Location-Python-master/README.md  
   creating: IP2Location-Python-master/data/
  inflating: IP2Location-Python-master/data/IP-COUNTRY.BIN  
  inflating: IP2Location-Python-master/data/IP2LOCATION-LITE-DB1.BIN  
  inflating: IP2Location-Python-master/data/IP2LOCATION-LITE-DB1.IPV6.BIN  
  inflating: IP2Location-Python-master/data/IPV6-COUNTRY.BIN  
  inflating: IP2Location-Python-master/data/country_test_ipv4_data.txt  
  inflating: IP2Location-Python-master/data/country_test_ipv6_data.txt  
  inflating: IP2Location-Python-master/lookup.py  
  inflating: IP2Location-Python-master/sample.py  
  inf

In [ ]:
!pip install IP2Location

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
# test example
#rec = database.get_all("89.108.114.66")

#print("Country Code          : " + rec.country_short)
#print("Country Name          : " + rec.country_long)
#print("Region Name           : " + rec.region)
#print("City Name             : " + rec.city)
#print("ISP Name              : " + rec.isp)
#print("Latitude              : " + rec.latitude)
#print("Longitude             : " + rec.longitude)
#print("Domain Name           : " + rec.domain)

Country Code          : RU
Country Name          : Russian Federation
Region Name           : This parameter is unavailable in selected .BIN data file. Please upgrade data file.
City Name             : This parameter is unavailable in selected .BIN data file. Please upgrade data file.
ISP Name              : This parameter is unavailable in selected .BIN data file. Please upgrade data file.
Latitude              : This parameter is unavailable in selected .BIN data file. Please upgrade data file.
Longitude             : This parameter is unavailable in selected .BIN data file. Please upgrade data file.
Domain Name           : This parameter is unavailable in selected .BIN data file. Please upgrade data file.


In [ ]:
import IP2Location
import ipinfo
from util import access_token_ipinfo


database = IP2Location.IP2Location("./IP2Location-Python-master/data/IP-COUNTRY.BIN")

class Country:

  def __init__(self, ip_address: str):
    self.ip_address = ip_address
    
  def get_country_ipinfo(self) -> str:
    handler = ipinfo.getHandler(access_token_ipinfo)
    details = handler.getDetails(self.ip_address)
    hostname = None
    if 'hostname' in details.all:
      hostname = details.hostname
    return details.country_name, hostname
    
  def get_country_ip2location(self) -> str:
    country = database.get_country_long(self.ip_address)
    # print(country, " ip2location")
    return country
  
  def get_country(self, hostname=False) -> str:
    if hostname:
      country_name, hostname = self.get_country_ipinfo()
      if country_name is None:
        country_name = self.get_country_ip2location()
      return country_name, hostname

    else:
      country_name = self.get_country_ip2location()
      if country_name == "INVALID IP ADDRESS":
        return None
      elif country_name.startswith("This parameter is unavailable") or country_name.startswith("-"):
        # print("redirect to ipinfo")
        country_name, _ = self.get_country_ipinfo()
      return country_name

In [ ]:
from tqdm import tqdm


countries = []
isps = []
hostnames = []

unavailable_sites = []  # this needs to be cleaned out from the dataset later

for entry in tqdm(concat_ngos.iloc[6603:].itertuples(), total=concat_ngos.shape[0]):  # .iloc[:50]

  # Get IP address
  try:
    ip_address = socket.gethostbyname(entry.website)
  except:
    unavailable_sites.append(entry.website)
    isps.append(None)
    countries.append(None)
    hostnames.append(None)
    continue

  country_handler = Country(ip_address)

  # Get ISP (provider) if possible
  if entry.bin_ssl == 1.0:
    isp_request = !curl {"http://ip-api.com/json/" + ip_address}
    try:
      response = json.loads(isp_request[0])
    except IndexError:
      print("weird Index Error")
      print(entry.website, response)
      response = json.loads(str(isp_request))

    #print(entry.website, response) # sometimes reposnse = []
    if not response:
      isps.append(None)
      countries.append(None)
      hostnames.append(None)
      continue

    if response["status"] == "success":
      isp = response["isp"]
      isps.append(isp)

      country = response["country"]
      countries.append(country)
      _, hostname = country_handler.get_country(hostname=True)
      hostnames.append(hostname)

    else:
      isps.append(None)
      country, hostname = country_handler.get_country(hostname=True)
      countries.append(country)
      hostnames.append(hostname)
  else:
    isps.append(None)
    country, hostname = country_handler.get_country(hostname=True)
    countries.append(country)
    hostnames.append(hostname)
 

  2%|▏         | 153/9545 [00:59<1:09:03,  2.27it/s]

weird Index Error
green-zoom.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '', 'lat': 59.9311, 'lon': 30.3609, 'timezone': 'Europe/Moscow', 'isp': 'OOO "Network of data-centers "Selectel"', 'org': 'Selectel Ltd.', 'as': 'AS49505 OOO Network of data-centers Selectel', 'query': '78.155.206.69'}


  2%|▏         | 154/9545 [00:59<56:56,  2.75it/s]  

weird Index Error
samaranordic.com []


  2%|▏         | 158/9545 [01:00<48:20,  3.24it/s]

weird Index Error
nochinadvolgoy.ru []


  3%|▎         | 310/9545 [01:57<55:26,  2.78it/s]

weird Index Error
cysticfibrosis.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190805', 'lat': 59.9391, 'lon': 30.2407, 'timezone': 'Europe/Moscow', 'isp': 'Virtual hosting BEGET.RU', 'org': 'Beget Ltd', 'as': 'AS198610 Beget LLC', 'query': '5.101.152.54'}


  3%|▎         | 311/9545 [01:57<51:58,  2.96it/s]

weird Index Error
sportclub88.ru []


  3%|▎         | 313/9545 [01:58<54:39,  2.82it/s]

weird Index Error
bel-roza.ru []


  3%|▎         | 318/9545 [02:00<53:29,  2.88it/s]

weird Index Error
fondmost.org {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '143441', 'lat': 55.8338, 'lon': 37.4025, 'timezone': 'Europe/Moscow', 'isp': '"Domain names registrar REG.RU", Ltd', 'org': 'Reg.Ru', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '31.31.196.189'}


  3%|▎         | 319/9545 [02:00<57:25,  2.68it/s]

weird Index Error
als-info.ru []


  3%|▎         | 320/9545 [02:01<53:05,  2.90it/s]

weird Index Error
give-world.com []


  3%|▎         | 321/9545 [02:01<57:07,  2.69it/s]

weird Index Error
dobrodel-kuban.ru []


  5%|▌         | 478/9545 [02:58<47:35,  3.17it/s]

weird Index Error
otirk.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '192241', 'lat': 59.9394, 'lon': 30.2283, 'timezone': 'Europe/Moscow', 'isp': 'BEGET.RU', 'org': 'Beget Ltd', 'as': 'AS198610 Beget LLC', 'query': '87.236.16.2'}


  5%|▌         | 479/9545 [02:59<50:52,  2.97it/s]

weird Index Error
ocean-msu.ru []


  5%|▌         | 486/9545 [03:01<49:49,  3.03it/s]

weird Index Error
scps-nsk.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '192241', 'lat': 59.9394, 'lon': 30.2283, 'timezone': 'Europe/Moscow', 'isp': 'BEGET.RU', 'org': 'Beget Ltd', 'as': 'AS198610 Beget LLC', 'query': '87.236.16.243'}


  7%|▋         | 628/9545 [04:03<57:11,  2.60it/s]

weird Index Error
shkolamil.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '', 'lat': 55.7778, 'lon': 37.5029, 'timezone': 'Europe/Moscow', 'isp': 'JSC "RU-CENTER"', 'org': 'JSC "RU-CENTER"', 'as': 'AS48287 JSC RU-CENTER', 'query': '194.85.61.76'}


  7%|▋         | 631/9545 [04:04<53:33,  2.77it/s]

weird Index Error
teplo-serdets.ru {'status': 'success', 'country': 'United States', 'countryCode': 'US', 'region': 'VA', 'regionName': 'Virginia', 'city': 'Ashburn', 'zip': '20147', 'lat': 39.018, 'lon': -77.539, 'timezone': 'America/New_York', 'isp': 'wix com inc', 'org': 'Wix Com Inc', 'as': 'AS58182 Wix.com Ltd.', 'query': '185.230.63.107'}


  8%|▊         | 783/9545 [05:02<46:09,  3.16it/s]

weird Index Error
intiu.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190005', 'lat': 59.8836, 'lon': 30.3303, 'timezone': 'Europe/Moscow', 'isp': 'SpaceWeb Spb', 'org': 'SpaceWeb CJSC', 'as': 'AS44112 SpaceWeb Ltd', 'query': '77.222.62.31'}


  8%|▊         | 785/9545 [05:03<42:27,  3.44it/s]

weird Index Error
nti2035.ru {'status': 'success', 'country': 'Canada', 'countryCode': 'CA', 'region': 'ON', 'regionName': 'Ontario', 'city': 'Toronto', 'zip': 'M5A', 'lat': 43.6532, 'lon': -79.3832, 'timezone': 'America/Toronto', 'isp': 'Cloudflare, Inc.', 'org': 'Cloudflare, Inc.', 'as': 'AS13335 Cloudflare, Inc.', 'query': '172.67.141.11'}


  8%|▊         | 788/9545 [05:04<44:39,  3.27it/s]

weird Index Error
opora82.ru []


  8%|▊         | 789/9545 [05:04<44:06,  3.31it/s]

weird Index Error
lubi-ludey.ru []


  8%|▊         | 791/9545 [05:05<43:22,  3.36it/s]

weird Index Error
fondgs.ru []


  8%|▊         | 793/9545 [05:05<43:15,  3.37it/s]

weird Index Error
ppouvng.ru []


  8%|▊         | 794/9545 [05:05<41:22,  3.53it/s]

weird Index Error
sknastroenie.ru []


  8%|▊         | 795/9545 [05:06<42:04,  3.47it/s]

weird Index Error
fondexpert.ru []


  8%|▊         | 796/9545 [05:06<42:02,  3.47it/s]

weird Index Error
atomsforhumanity.ru []


 10%|▉         | 919/9545 [05:53<53:37,  2.68it/s]

weird Index Error
myrostrf.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOS', 'regionName': 'Moscow Oblast', 'city': 'Skolkovo', 'zip': '143025', 'lat': 55.6916, 'lon': 37.3474, 'timezone': 'Europe/Moscow', 'isp': 'JSC IOT', 'org': 'JSC IOT', 'as': 'AS29182 JSC IOT', 'query': '188.120.254.71'}


 10%|▉         | 921/9545 [05:54<56:47,  2.53it/s]

weird Index Error
cxt-videokniga.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'Reg.Ru', 'org': '', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '37.140.192.13'}


 10%|▉         | 923/9545 [05:54<54:03,  2.66it/s]

weird Index Error
crimea-moscow.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'IRK', 'regionName': 'Irkutsk Oblast', 'city': 'Irkutsk', 'zip': '664000', 'lat': 52.2855, 'lon': 104.289, 'timezone': 'Asia/Irkutsk', 'isp': 'TheServer', 'org': 'JSC Server', 'as': 'AS29182 JSC IOT', 'query': '185.43.7.111'}


 10%|▉         | 926/9545 [05:55<49:10,  2.92it/s]

weird Index Error
malyshbf.ru []


 10%|▉         | 927/9545 [05:56<42:22,  3.39it/s]

weird Index Error
rehabprofi.ru []


 10%|▉         | 929/9545 [05:56<39:43,  3.61it/s]

weird Index Error
fedordobronravov.ru []


 10%|▉         | 933/9545 [05:59<1:30:16,  1.59it/s]

weird Index Error
нроо.рф {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'Variti LLC', 'org': 'Tilda', 'as': 'AS64432 Variti LLC', 'query': '185.165.123.36'}


 10%|▉         | 934/9545 [05:59<1:20:13,  1.79it/s]

weird Index Error
silverexpress.ru []


 10%|▉         | 935/9545 [06:00<1:08:18,  2.10it/s]

weird Index Error
xoroshix-nko.ru []


 10%|▉         | 936/9545 [06:00<1:00:26,  2.37it/s]

weird Index Error
zabota-i-ukhod.ru []


 10%|▉         | 937/9545 [06:00<54:39,  2.62it/s]  

weird Index Error
моясемья34.рф []


 10%|▉         | 939/9545 [06:02<1:23:23,  1.72it/s]

weird Index Error
sdomado.ru []


 10%|▉         | 942/9545 [06:03<58:05,  2.47it/s]  

weird Index Error
www.allianceforanimals.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'TA', 'regionName': 'Tatarstan Republic', 'city': 'Kazan’', 'zip': '420000', 'lat': 55.785, 'lon': 49.1178, 'timezone': 'Europe/Moscow', 'isp': 'SELECTEL', 'org': '', 'as': 'AS50340 OOO Network of data-centers Selectel', 'query': '185.17.10.59'}


 10%|▉         | 944/9545 [06:03<52:28,  2.73it/s]

weird Index Error
fedskate.ru []


 10%|▉         | 947/9545 [06:04<47:53,  2.99it/s]

weird Index Error
fsrbit.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '', 'lat': 55.7029, 'lon': 37.6968, 'timezone': 'Europe/Moscow', 'isp': '.masterhost', 'org': '', 'as': 'AS25532 LLC MASTERHOST', 'query': '90.156.201.28'}


 11%|█▏        | 1084/9545 [07:01<49:49,  2.83it/s]

weird Index Error
turgeneff.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '143441', 'lat': 55.8338, 'lon': 37.4025, 'timezone': 'Europe/Moscow', 'isp': '"Domain names registrar REG.RU", Ltd', 'org': 'Reg.Ru', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '31.31.196.247'}


 11%|█▏        | 1085/9545 [07:02<47:08,  2.99it/s]

weird Index Error
manepa.ru []


 11%|█▏        | 1086/9545 [07:02<45:16,  3.11it/s]

weird Index Error
лучнадежды.рф []


 11%|█▏        | 1087/9545 [07:02<48:24,  2.91it/s]

weird Index Error
almetseasons.ru []


 11%|█▏        | 1088/9545 [07:03<50:26,  2.79it/s]

weird Index Error
www.lightinhands.ru []


 11%|█▏        | 1089/9545 [07:03<47:49,  2.95it/s]

weird Index Error
profsobranie.ru []


 11%|█▏        | 1090/9545 [07:03<49:02,  2.87it/s]

weird Index Error
arpot.ru []


 11%|█▏        | 1092/9545 [07:05<1:23:23,  1.69it/s]

weird Index Error
export67.ru []


 11%|█▏        | 1093/9545 [07:05<1:11:02,  1.98it/s]

weird Index Error
fond.elitsy.ru []


 11%|█▏        | 1094/9545 [07:06<1:04:16,  2.19it/s]

weird Index Error
crkdomnadejd.ru []


 11%|█▏        | 1097/9545 [07:07<55:38,  2.53it/s]

weird Index Error
project-initiative.ru []


 12%|█▏        | 1099/9545 [07:07<48:15,  2.92it/s]

weird Index Error
onmywings.ru []


 12%|█▏        | 1102/9545 [07:08<41:50,  3.36it/s]

weird Index Error
sirotstvynet.ru []


 12%|█▏        | 1106/9545 [07:10<51:50,  2.71it/s]

weird Index Error
franchise-norilsk.ru {'status': 'success', 'country': 'Canada', 'countryCode': 'CA', 'region': 'ON', 'regionName': 'Ontario', 'city': 'Toronto', 'zip': 'M5A', 'lat': 43.6532, 'lon': -79.3832, 'timezone': 'America/Toronto', 'isp': 'Cloudflare, Inc.', 'org': 'Cloudflare, Inc.', 'as': 'AS13335 Cloudflare, Inc.', 'query': '172.67.165.84'}


 12%|█▏        | 1107/9545 [07:11<48:35,  2.89it/s]

weird Index Error
intercurs.ru []


 13%|█▎        | 1243/9545 [08:06<54:08,  2.56it/s]

weird Index Error
rccpenza.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '143441', 'lat': 55.8338, 'lon': 37.4025, 'timezone': 'Europe/Moscow', 'isp': '"Domain names registrar REG.RU", Ltd', 'org': 'Reg.Ru', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '31.31.196.73'}


 13%|█▎        | 1246/9545 [08:08<1:29:32,  1.54it/s]

weird Index Error
epona-hippotherapy.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOS', 'regionName': 'Moscow Oblast', 'city': 'Skolkovo', 'zip': '143025', 'lat': 55.6916, 'lon': 37.3474, 'timezone': 'Europe/Moscow', 'isp': 'JSC IOT', 'org': 'JSC IOT', 'as': 'AS29182 JSC IOT', 'query': '94.250.250.97'}


 13%|█▎        | 1247/9545 [08:08<1:17:03,  1.79it/s]

weird Index Error
dismas.ru []


 15%|█▍        | 1408/9545 [09:06<45:21,  2.99it/s]

weird Index Error
vestaplus86.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'MCHOST', 'org': '', 'as': 'AS210079 EuroByte LLC', 'query': '95.142.38.59'}


 15%|█▍        | 1409/9545 [09:07<42:21,  3.20it/s]

weird Index Error
живуспортом.рф []


 15%|█▍        | 1410/9545 [09:07<41:58,  3.23it/s]

weird Index Error
hr-insider.ru []


 15%|█▍        | 1412/9545 [09:08<39:16,  3.45it/s]

weird Index Error
ano-nadezhda.ru []


 15%|█▍        | 1413/9545 [09:08<45:08,  3.00it/s]

weird Index Error
chyltys.ru []


 15%|█▍        | 1414/9545 [09:09<53:09,  2.55it/s]

weird Index Error
dev-expoolimp.ru []


 15%|█▍        | 1415/9545 [09:09<48:45,  2.78it/s]

weird Index Error
kuzfond.ru []


 15%|█▍        | 1416/9545 [09:09<41:27,  3.27it/s]

weird Index Error
immasters.ru []


 15%|█▍        | 1417/9545 [09:09<43:24,  3.12it/s]

weird Index Error
www.plusr.ru []


 15%|█▍        | 1418/9545 [09:10<42:17,  3.20it/s]

weird Index Error
beliy-parohod.ru []


 15%|█▍        | 1420/9545 [09:11<54:21,  2.49it/s]

weird Index Error
soc38.ru []


 15%|█▍        | 1422/9545 [09:11<49:16,  2.75it/s]

weird Index Error
batalistneva.ru []


 15%|█▍        | 1424/9545 [09:12<50:44,  2.67it/s]

weird Index Error
gm-inginirium.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'Servers.com, Inc.', 'org': 'Servers RU Mow1', 'as': 'AS39134 United Network LLC', 'query': '88.212.247.52'}


 15%|█▍        | 1425/9545 [09:12<47:28,  2.85it/s]

weird Index Error
rosgonki.ru []


 17%|█▋        | 1575/9545 [10:08<41:46,  3.18it/s]

weird Index Error
fcasi.ru {'status': 'success', 'country': 'United States', 'countryCode': 'US', 'region': 'VA', 'regionName': 'Virginia', 'city': 'Ashburn', 'zip': '20147', 'lat': 39.018, 'lon': -77.539, 'timezone': 'America/New_York', 'isp': 'wix com inc', 'org': 'Wix Com Inc', 'as': 'AS58182 Wix.com Ltd.', 'query': '185.230.63.171'}


 17%|█▋        | 1576/9545 [10:09<40:39,  3.27it/s]

weird Index Error
sofund.ru []


 17%|█▋        | 1578/9545 [10:09<45:55,  2.89it/s]

weird Index Error
svgb71.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '192241', 'lat': 59.9394, 'lon': 30.2283, 'timezone': 'Europe/Moscow', 'isp': 'BEGET.RU', 'org': 'Beget Ltd', 'as': 'AS198610 Beget LLC', 'query': '87.236.16.254'}


 17%|█▋        | 1582/9545 [10:11<39:20,  3.37it/s]

weird Index Error
fly-slon.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '143441', 'lat': 55.84, 'lon': 37.4069, 'timezone': 'Europe/Moscow', 'isp': '"Domain names registrar REG.RU", Ltd', 'org': 'Reg.Ru', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '31.31.198.173'}


 17%|█▋        | 1583/9545 [10:11<34:45,  3.82it/s]

weird Index Error
фондсодружество.рф []


 17%|█▋        | 1590/9545 [10:14<1:10:14,  1.89it/s]

weird Index Error
ulrgo.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '', 'lat': 59.9345, 'lon': 30.4357, 'timezone': 'Europe/Moscow', 'isp': 'BEGET.RU', 'org': '', 'as': 'AS198610 Beget LLC', 'query': '45.130.41.33'}


 18%|█▊        | 1705/9545 [11:06<58:38,  2.23it/s]  

weird Index Error
бенди-татарстан.рф {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'LEN', 'regionName': 'Leningrad Oblast', 'city': 'Untolovo', 'zip': '', 'lat': 59.9828, 'lon': 30.2133, 'timezone': 'Europe/Moscow', 'isp': 'Sprinthost P17', 'org': '', 'as': 'AS35278 SPRINTHOST.RU LLC', 'query': '141.8.195.205'}


 18%|█▊        | 1706/9545 [11:06<47:23,  2.76it/s]

weird Index Error
openyourhearts.org []


 18%|█▊        | 1707/9545 [11:06<44:09,  2.96it/s]

weird Index Error
standart-center.ru []


 18%|█▊        | 1709/9545 [11:07<42:44,  3.06it/s]

weird Index Error
educ-center.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'RYA', 'regionName': 'Ryazan Oblast', 'city': 'Ryazan', 'zip': '390000', 'lat': 54.613, 'lon': 39.743, 'timezone': 'Europe/Moscow', 'isp': 'Megagroup.ru LLC', 'org': 'Megagroup.ru LLC', 'as': 'AS60357 Megagroup.ru LLC', 'query': '185.32.58.255'}


 18%|█▊        | 1715/9545 [11:09<48:24,  2.70it/s]

weird Index Error
redcross-auction.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'LEN', 'regionName': 'Leningrad Oblast', 'city': 'Untolovo', 'zip': '', 'lat': 59.9828, 'lon': 30.2133, 'timezone': 'Europe/Moscow', 'isp': 'Sprinthost P17', 'org': '', 'as': 'AS35278 SPRINTHOST.RU LLC', 'query': '141.8.195.33'}


 18%|█▊        | 1719/9545 [11:11<47:12,  2.76it/s]

weird Index Error
trfl.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '101194', 'lat': 55.7332, 'lon': 37.5833, 'timezone': 'Europe/Moscow', 'isp': 'Yandex enterprise network', 'org': 'Yandex LLC', 'as': 'AS200350 Yandex.Cloud LLC', 'query': '84.201.137.135'}


 18%|█▊        | 1721/9545 [11:12<1:09:33,  1.87it/s]

weird Index Error
nko-maximum.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '101194', 'lat': 55.7332, 'lon': 37.5833, 'timezone': 'Europe/Moscow', 'isp': 'Yandex enterprise network', 'org': 'Yandex LLC', 'as': 'AS200350 Yandex.Cloud LLC', 'query': '84.201.137.135'}


 18%|█▊        | 1722/9545 [11:13<1:05:22,  1.99it/s]

weird Index Error
russianspaniel.ru []


 18%|█▊        | 1724/9545 [11:13<51:58,  2.51it/s]

weird Index Error
соцработник48.рф []


 18%|█▊        | 1725/9545 [11:14<55:15,  2.36it/s]

weird Index Error
dobroye-delo.ru []


 18%|█▊        | 1731/9545 [11:16<48:15,  2.70it/s]

weird Index Error
ano-cspsaulyk.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '', 'lat': 59.9345, 'lon': 30.4357, 'timezone': 'Europe/Moscow', 'isp': 'BEGET.RU', 'org': '', 'as': 'AS198610 Beget LLC', 'query': '45.130.41.9'}


 19%|█▉        | 1848/9545 [12:03<50:48,  2.53it/s]

weird Index Error
oppozs.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190805', 'lat': 59.9417, 'lon': 30.3096, 'timezone': 'Europe/Moscow', 'isp': 'TimeWeb Ltd.', 'org': 'TimeWeb Ltd', 'as': 'AS9123 TimeWeb Ltd.', 'query': '185.114.245.107'}


 19%|█▉        | 1854/9545 [12:05<48:31,  2.64it/s]

weird Index Error
landoflife.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '192241', 'lat': 59.9394, 'lon': 30.2283, 'timezone': 'Europe/Moscow', 'isp': 'BEGET.RU', 'org': 'Beget Ltd', 'as': 'AS198610 Beget LLC', 'query': '87.236.16.222'}


 19%|█▉        | 1856/9545 [12:06<44:26,  2.88it/s]

weird Index Error
молцентрбрянск.рф {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '', 'lat': 59.9345, 'lon': 30.4357, 'timezone': 'Europe/Moscow', 'isp': 'BEGET.RU', 'org': '', 'as': 'AS198610 Beget LLC', 'query': '45.130.41.6'}


 19%|█▉        | 1857/9545 [12:06<41:49,  3.06it/s]

weird Index Error
peer-partners.ru []


 19%|█▉        | 1858/9545 [12:07<40:06,  3.19it/s]

weird Index Error
my-krymchane.ru []


 19%|█▉        | 1859/9545 [12:07<38:42,  3.31it/s]

weird Index Error
dsluchik.ru []


 19%|█▉        | 1860/9545 [12:07<43:29,  2.95it/s]

weird Index Error
pantheum.ru []


 19%|█▉        | 1861/9545 [12:08<45:28,  2.82it/s]

weird Index Error
зашкольноепитание.рф []


 20%|█▉        | 1862/9545 [12:08<42:25,  3.02it/s]

weird Index Error
rvio61.ru []


 20%|█▉        | 1863/9545 [12:08<45:56,  2.79it/s]

weird Index Error
просто-живи.рф []


 20%|█▉        | 1864/9545 [12:09<41:30,  3.08it/s]

weird Index Error
priyut67.ru []


 20%|█▉        | 1865/9545 [12:09<35:45,  3.58it/s]

weird Index Error
wncf.ru []


 21%|██        | 2002/9545 [13:16<46:31,  2.70it/s]

weird Index Error
pro-raion.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '143441', 'lat': 55.84, 'lon': 37.4069, 'timezone': 'Europe/Moscow', 'isp': '"Domain names registrar REG.RU", Ltd', 'org': 'Reg.Ru', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '31.31.198.44'}


 22%|██▏       | 2145/9545 [14:18<43:42,  2.82it/s]

weird Index Error
mfoprim.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '143441', 'lat': 55.84, 'lon': 37.4069, 'timezone': 'Europe/Moscow', 'isp': '"Domain names registrar REG.RU", Ltd', 'org': 'Reg.Ru', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '31.31.198.39'}


 22%|██▏       | 2146/9545 [14:18<44:08,  2.79it/s]

weird Index Error
просодействие.рф []


 23%|██▎       | 2149/9545 [14:19<40:07,  3.07it/s]

weird Index Error
mb-62.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'RYA', 'regionName': 'Ryazan Oblast', 'city': 'Ryazan', 'zip': '390005', 'lat': 54.6161, 'lon': 39.7376, 'timezone': 'Europe/Moscow', 'isp': 'OnTelecom LLC', 'org': 'OnTelecom LLC', 'as': 'AS60042 OnTelecom LLC', 'query': '176.118.221.7'}


 23%|██▎       | 2156/9545 [14:21<45:29,  2.71it/s]

weird Index Error
stranatvortsov.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '', 'lat': 55.7972, 'lon': 37.5656, 'timezone': 'Europe/Moscow', 'isp': 'RTCOMM', 'org': 'Avguro Technologies Ltd.', 'as': 'AS8342 JSC RTComm.RU', 'query': '81.177.141.224'}


 24%|██▍       | 2270/9545 [15:09<46:20,  2.62it/s]

weird Index Error
sport-pokolenie.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190005', 'lat': 59.8889, 'lon': 30.3296, 'timezone': 'Europe/Moscow', 'isp': 'TIMEWEB Co Ltd.', 'org': 'TIMEWEB', 'as': 'AS9123 TimeWeb Ltd.', 'query': '5.23.51.236'}


 24%|██▍       | 2272/9545 [15:09<33:58,  3.57it/s]

weird Index Error
наши-дети.org []


 24%|██▍       | 2274/9545 [15:10<42:05,  2.88it/s]

weird Index Error
torya.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '', 'lat': 55.7215, 'lon': 37.6968, 'timezone': 'Europe/Moscow', 'isp': 'Ddos-guard LTD', 'org': 'Tilda Publishing LLC', 'as': 'AS57724 DDOS-GUARD LTD', 'query': '185.215.4.10'}


 24%|██▍       | 2275/9545 [15:10<39:50,  3.04it/s]

weird Index Error
manufactura-mi.ru []


 24%|██▍       | 2276/9545 [15:11<38:03,  3.18it/s]

weird Index Error
toysmk.ru []


 24%|██▍       | 2277/9545 [15:11<40:32,  2.99it/s]

weird Index Error
fond770.ru []


 24%|██▍       | 2281/9545 [15:12<35:21,  3.42it/s]

weird Index Error
yadlaries.ru []


 24%|██▍       | 2284/9545 [15:14<50:37,  2.39it/s]

weird Index Error
startum-dpo.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'OOO "Network of data-centers "Selectel"', 'org': 'Selectel', 'as': 'AS50340 OOO Network of data-centers Selectel', 'query': '82.148.18.17'}


 24%|██▍       | 2289/9545 [15:15<35:52,  3.37it/s]

weird Index Error
uramkb.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '', 'lat': 55.7778, 'lon': 37.5029, 'timezone': 'Europe/Moscow', 'isp': 'JSC "RU-CENTER"', 'org': 'HostingCenter router', 'as': 'AS48287 JSC RU-CENTER', 'query': '91.189.114.17'}


 24%|██▍       | 2293/9545 [15:17<36:45,  3.29it/s]

weird Index Error
soc-help.ru []


 24%|██▍       | 2295/9545 [15:17<39:52,  3.03it/s]

weird Index Error
fondsg.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190805', 'lat': 59.9417, 'lon': 30.3096, 'timezone': 'Europe/Moscow', 'isp': 'TimeWeb Ltd.', 'org': 'TimeWeb Ltd', 'as': 'AS9123 TimeWeb Ltd.', 'query': '185.114.247.102'}


 24%|██▍       | 2296/9545 [15:18<40:53,  2.95it/s]

weird Index Error
ano-rsi.ru []


 24%|██▍       | 2297/9545 [15:18<42:37,  2.83it/s]

weird Index Error
vpolyota.ru []


 24%|██▍       | 2298/9545 [15:18<40:03,  3.02it/s]

weird Index Error
tvtambov.ru []


 24%|██▍       | 2299/9545 [15:19<37:50,  3.19it/s]

weird Index Error
pooofog.ru []


 24%|██▍       | 2301/9545 [15:20<59:09,  2.04it/s]  

weird Index Error
фонд-гиппокамп.рф []


 24%|██▍       | 2303/9545 [15:21<45:59,  2.62it/s]

weird Index Error
gorodarb.ru []


 24%|██▍       | 2304/9545 [15:21<45:30,  2.65it/s]

weird Index Error
irgrb.ru []


 24%|██▍       | 2305/9545 [15:21<43:08,  2.80it/s]

weird Index Error
bfsapfir.ru []


 24%|██▍       | 2306/9545 [15:22<40:39,  2.97it/s]

weird Index Error
www.dobro-mami.ru []


 24%|██▍       | 2307/9545 [15:22<38:48,  3.11it/s]

weird Index Error
harmonyhomes.ru []


 24%|██▍       | 2308/9545 [15:22<37:45,  3.19it/s]

weird Index Error
fund-veles.ru []


 24%|██▍       | 2309/9545 [15:23<41:27,  2.91it/s]

weird Index Error
profifuture.ru []


 24%|██▍       | 2311/9545 [15:23<43:56,  2.74it/s]

weird Index Error
myvserasnye.ru []


 25%|██▌       | 2422/9545 [16:12<45:52,  2.59it/s]

weird Index Error
vmestevpered.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'InSys', 'org': 'iHead company', 'as': 'AS42358 PJSC Rostelecom', 'query': '46.255.97.135'}


 25%|██▌       | 2425/9545 [16:13<47:08,  2.52it/s]

weird Index Error
murkosha.ru []


 25%|██▌       | 2428/9545 [16:14<46:59,  2.52it/s]

weird Index Error
nkofestival.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '121087', 'lat': 55.7483, 'lon': 37.6171, 'timezone': 'Europe/Moscow', 'isp': 'Compubyte Limited', 'org': 'Compubyte Limited', 'as': 'AS204343 Compubyte Limited', 'query': '193.109.246.9'}


 25%|██▌       | 2430/9545 [16:15<45:52,  2.58it/s]

weird Index Error
образованиеонлайн.рф {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '143441', 'lat': 55.84, 'lon': 37.4069, 'timezone': 'Europe/Moscow', 'isp': '"Domain names registrar REG.RU", Ltd', 'org': 'Reg.Ru', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '31.31.198.219'}


 26%|██▌       | 2434/9545 [16:16<37:24,  3.17it/s]

weird Index Error
sa-fond.ru []


 26%|██▌       | 2436/9545 [16:17<39:47,  2.98it/s]

weird Index Error
aamos.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'KEM', 'regionName': 'Kemerovo Oblast', 'city': 'Mezhdurechensk', 'zip': '652870', 'lat': 53.6958, 'lon': 88.0648, 'timezone': 'Asia/Novokuznetsk', 'isp': 'Inteks Service Ltd', 'org': 'Inteks-Service', 'as': 'AS39860 Inteks Service Ltd.', 'query': '213.110.60.166'}


 26%|██▌       | 2437/9545 [16:17<37:56,  3.12it/s]

weird Index Error
youngdesignspb.ru []


 26%|██▌       | 2443/9545 [16:20<43:04,  2.75it/s]

weird Index Error
www.opencityekb.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '', 'lat': 59.9311, 'lon': 30.3609, 'timezone': 'Europe/Moscow', 'isp': 'OOO "Network of data-centers "Selectel"', 'org': '', 'as': 'AS49505 OOO Network of data-centers Selectel', 'query': '89.248.193.151'}


 26%|██▌       | 2444/9545 [16:20<43:04,  2.75it/s]

weird Index Error
людиарктики.рф []


 26%|██▌       | 2446/9545 [16:21<38:41,  3.06it/s]

weird Index Error
ftim.ru []


 26%|██▌       | 2450/9545 [16:22<41:41,  2.84it/s]

weird Index Error
ano-orientir.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'Variti LLC', 'org': 'Tilda', 'as': 'AS64432 Variti LLC', 'query': '185.165.123.36'}


 26%|██▌       | 2451/9545 [16:22<40:27,  2.92it/s]

weird Index Error
50gkhecokontrol.ru []


 26%|██▌       | 2453/9545 [16:23<38:38,  3.06it/s]

weird Index Error
whitegeneral.ru []


 26%|██▌       | 2454/9545 [16:23<40:49,  2.89it/s]

weird Index Error
opora-help.ru []


 26%|██▌       | 2455/9545 [16:24<39:03,  3.03it/s]

weird Index Error
opora-news.ru []


 26%|██▌       | 2456/9545 [16:24<39:57,  2.96it/s]

weird Index Error
angelvery.ru []


 26%|██▌       | 2457/9545 [16:24<40:43,  2.90it/s]

weird Index Error
aap-pro.ru []


 27%|██▋       | 2585/9545 [17:15<40:26,  2.87it/s]

weird Index Error
ark-in.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190805', 'lat': 59.9417, 'lon': 30.3096, 'timezone': 'Europe/Moscow', 'isp': 'TimeWeb Ltd.', 'org': 'TimeWeb Ltd', 'as': 'AS9123 TimeWeb Ltd.', 'query': '185.114.245.109'}


 27%|██▋       | 2587/9545 [17:16<47:37,  2.43it/s]

weird Index Error
helpsocial.ru []


 27%|██▋       | 2588/9545 [17:16<51:25,  2.25it/s]

weird Index Error
krsk2019.ru []


 27%|██▋       | 2589/9545 [17:17<49:18,  2.35it/s]

weird Index Error
shkolapavlenko.ru []


 27%|██▋       | 2592/9545 [17:18<41:23,  2.80it/s]

weird Index Error
nocrb.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '142770', 'lat': 55.5913, 'lon': 37.5395, 'timezone': 'Europe/Moscow', 'isp': 'Okay-Telecom Ltd.', 'org': 'Okay telecom Voxility', 'as': 'AS199669 Okay-Telecom Ltd.', 'query': '185.10.44.2'}


 27%|██▋       | 2595/9545 [17:19<32:34,  3.56it/s]

weird Index Error
человекнапротив.рф []


 27%|██▋       | 2596/9545 [17:19<35:49,  3.23it/s]

weird Index Error
planetary-spb.ru []


 27%|██▋       | 2597/9545 [17:19<37:23,  3.10it/s]

weird Index Error
ocimt.ru []


 27%|██▋       | 2600/9545 [17:21<1:03:21,  1.83it/s]

weird Index Error
музейвоин.рф []


 27%|██▋       | 2601/9545 [17:22<57:02,  2.03it/s]  

weird Index Error
центрфинансирования.рф []


 27%|██▋       | 2602/9545 [17:22<45:38,  2.54it/s]

weird Index Error
spbsacad.ru []


 27%|██▋       | 2604/9545 [17:23<45:29,  2.54it/s]

weird Index Error
vremya-dobryh.ru []


 27%|██▋       | 2605/9545 [17:23<41:48,  2.77it/s]

weird Index Error
blagosfera7ya.ru []


 27%|██▋       | 2606/9545 [17:23<42:48,  2.70it/s]

weird Index Error
dva-kryla.ru []


 27%|██▋       | 2607/9545 [17:24<42:30,  2.72it/s]

weird Index Error
godnauki.ru []


 27%|██▋       | 2608/9545 [17:24<43:04,  2.68it/s]

weird Index Error
nationalpriority.ru []


 27%|██▋       | 2609/9545 [17:24<40:14,  2.87it/s]

weird Index Error
nationalprojects.ru []


 27%|██▋       | 2613/9545 [17:25<36:35,  3.16it/s]

weird Index Error
tyvigre.ru []


 27%|██▋       | 2615/9545 [17:26<41:41,  2.77it/s]

weird Index Error
годнауки.рф {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'LLC VK', 'org': 'VK Cloud Solutions', 'as': 'AS47764 LLC VK', 'query': '87.239.107.107'}


 29%|██▊       | 2730/9545 [18:09<48:00,  2.37it/s]

weird Index Error
спбну.рф {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190005', 'lat': 59.8889, 'lon': 30.3296, 'timezone': 'Europe/Moscow', 'isp': 'TimeWeb Ltd.', 'org': 'TimeWeb Ltd', 'as': 'AS9123 TimeWeb Ltd.', 'query': '109.68.214.217'}


 29%|██▊       | 2733/9545 [18:10<41:31,  2.73it/s]

weird Index Error
vertigo-project.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '', 'lat': 55.7215, 'lon': 37.6968, 'timezone': 'Europe/Moscow', 'isp': 'Ddos-guard LTD', 'org': 'Tilda Publishing LLC', 'as': 'AS57724 DDOS-GUARD LTD', 'query': '185.215.4.16'}


 29%|██▊       | 2734/9545 [18:10<38:48,  2.93it/s]

weird Index Error
delaydobro55.ru []


 29%|██▊       | 2735/9545 [18:11<39:01,  2.91it/s]

weird Index Error
papa-perm.ru []


 29%|██▊       | 2739/9545 [18:12<36:39,  3.09it/s]

weird Index Error
sds-shkola.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '125412', 'lat': 55.7483, 'lon': 37.6171, 'timezone': 'Europe/Moscow', 'isp': 'G-Core Labs S.A.', 'org': 'G-Core Labs S.A', 'as': 'AS199524 G-Core Labs S.A.', 'query': '83.229.25.22'}


 29%|██▊       | 2741/9545 [18:13<39:28,  2.87it/s]

weird Index Error
baltachlar.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190805', 'lat': 59.9417, 'lon': 30.3096, 'timezone': 'Europe/Moscow', 'isp': 'TimeWeb Ltd.', 'org': 'TimeWeb Ltd', 'as': 'AS9123 TimeWeb Ltd.', 'query': '185.114.245.100'}


 29%|██▊       | 2743/9545 [18:13<40:02,  2.83it/s]

weird Index Error
mosmed-support.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'Reg.Ru', 'org': '', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '37.140.192.63'}


 29%|██▉       | 2749/9545 [18:16<41:28,  2.73it/s]

weird Index Error
neirodety.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '143441', 'lat': 55.8338, 'lon': 37.4025, 'timezone': 'Europe/Moscow', 'isp': '"Domain names registrar REG.RU", Ltd', 'org': 'Reg.Ru', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '31.31.196.170'}


 29%|██▉       | 2755/9545 [18:18<34:10,  3.31it/s]

weird Index Error
dogchance.tilda.ws {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190005', 'lat': 59.8886, 'lon': 30.3287, 'timezone': 'Europe/Moscow', 'isp': 'TimeWeb Co. LTD', 'org': 'TimeWeb Ltd.', 'as': 'AS9123 TimeWeb Ltd.', 'query': '176.57.210.144'}


 29%|██▉       | 2756/9545 [18:18<40:08,  2.82it/s]

weird Index Error
infradet.ru []


 29%|██▉       | 2758/9545 [18:19<36:39,  3.09it/s]

weird Index Error
diritc.ru []


 29%|██▉       | 2759/9545 [18:19<38:33,  2.93it/s]

weird Index Error
дмтк.рф []


 29%|██▉       | 2760/9545 [18:20<41:28,  2.73it/s]

weird Index Error
acro-dance.ru []


 29%|██▉       | 2762/9545 [18:20<38:33,  2.93it/s]

weird Index Error
preobrazhenskyil.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '192241', 'lat': 59.9394, 'lon': 30.2283, 'timezone': 'Europe/Moscow', 'isp': 'BEGET.RU', 'org': 'Beget Ltd', 'as': 'AS198610 Beget LLC', 'query': '87.236.16.19'}


 29%|██▉       | 2763/9545 [18:22<1:11:36,  1.58it/s]

weird Index Error
юнонапраздник.рф []


 29%|██▉       | 2768/9545 [18:23<46:56,  2.41it/s]

weird Index Error
анодполидер.рф {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'Ddos-guard LTD', 'org': 'TildaPublishing', 'as': 'AS57724 DDOS-GUARD LTD', 'query': '185.129.100.112'}


 29%|██▉       | 2770/9545 [18:24<45:41,  2.47it/s]

weird Index Error
fond214ur.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOS', 'regionName': 'Moscow Oblast', 'city': 'Skolkovo', 'zip': '143025', 'lat': 55.6916, 'lon': 37.3474, 'timezone': 'Europe/Moscow', 'isp': 'JSC IOT', 'org': 'JSC IOT', 'as': 'AS29182 JSC IOT', 'query': '79.143.72.235'}


 29%|██▉       | 2773/9545 [18:25<43:07,  2.62it/s]

weird Index Error
art-teach.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'Reg.Ru', 'org': '', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '37.140.192.24'}


 29%|██▉       | 2776/9545 [18:26<36:50,  3.06it/s]

weird Index Error
informpskov.ru []


 30%|███       | 2908/9545 [19:23<38:44,  2.86it/s]

weird Index Error
vorotavsport.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'SPE', 'regionName': 'St.-Petersburg', 'city': 'St Petersburg', 'zip': '190005', 'lat': 59.8889, 'lon': 30.3296, 'timezone': 'Europe/Moscow', 'isp': 'TimeWeb Co. LTD', 'org': '', 'as': 'AS9123 TimeWeb Ltd.', 'query': '92.53.96.4'}


 31%|███       | 2913/9545 [19:25<37:11,  2.97it/s]

weird Index Error
цпв-гвардия.рф {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'OOO "Network of data-centers "Selectel"', 'org': 'Selectel', 'as': 'AS49505 OOO Network of data-centers Selectel', 'query': '185.137.235.2'}


 31%|███       | 2914/9545 [19:25<35:38,  3.10it/s]

weird Index Error
anorsi.ru []


 31%|███       | 2917/9545 [19:26<31:49,  3.47it/s]

weird Index Error
academiasport76.ru {'status': 'success', 'country': 'Russia', 'countryCode': 'RU', 'region': 'MOW', 'regionName': 'Moscow', 'city': 'Moscow', 'zip': '144700', 'lat': 55.7558, 'lon': 37.6173, 'timezone': 'Europe/Moscow', 'isp': 'Reg.Ru', 'org': '', 'as': 'AS197695 Domain names registrar REG.RU, Ltd', 'query': '37.140.192.209'}


 31%|███       | 2918/9545 [19:27<37:41,  2.93it/s]

weird Index Error
agroturizm54.ru []


 31%|███       | 2919/9545 [19:27<39:11,  2.82it/s]

weird Index Error
ipipm.ru []


 31%|███       | 2921/9545 [19:28<37:48,  2.92it/s]

weird Index Error
createdu.ru []


 31%|███       | 2923/9545 [19:29<52:08,  2.12it/s]  

weird Index Error
бфн.рф []


 31%|███       | 2924/9545 [19:29<49:18,  2.24it/s]

weird Index Error
rck51.ru []


 31%|███       | 2942/9545 [19:38<44:04,  2.50it/s]


In [ ]:
# check
len(countries), len(hostnames), len(isps)

(2942, 2942, 2942)

In [ ]:
# for the first 2566
# concat_ngos['website_country'] = countries + [None] * (len(concat_ngos) - 2566)
# concat_ngos['website_hostname'] = hostnames + [None] * (len(concat_ngos) - 2566)
# concat_ngos['website_isp'] = isps[:-1] + [None] * (len(concat_ngos) - 2566)

In [ ]:
for i in range(len(hostnames)):
  concat_ngos.at[i+6603, 'website_country'] = countries[i]
  concat_ngos.at[i+6603, 'website_hostname'] = hostnames[i]
  concat_ngos.at[i+6603, 'website_isp'] = isps[i]

In [ ]:
2566+4037

6603

In [ ]:
concat_ngos.iloc[9540:]

ogrn          website  bin_robots  bin_sitemap  bin_fb  bin_vk  \
9540  1217700385539    ngolikeyou.ru         1.0          1.0     1.0     1.0   
9541  1214600009524         kcmol.ru         1.0          0.0     0.0     1.0   
9542  1217700396330       gorneks.ru         1.0          1.0     0.0     0.0   
9543  1213300006920   ano-nachalo.ru         1.0          0.0     0.0     0.0   
9544  1212300052249  blagfond-zzh.ru         0.0          1.0     0.0     0.0   

      bin_ig  bin_ok  bin_youtube  bin_tiktok  ...  \
9540     1.0     0.0          1.0         0.0  ...   
9541     1.0     0.0          0.0         0.0  ...   
9542     0.0     0.0          0.0         0.0  ...   
9543     0.0     0.0          0.0         0.0  ...   
9544     0.0     0.0          0.0         0.0  ...   

                                              shortName     regDate  \
9540  АНО ЦЕНТР СИСТЕМНОЙ ПОДДЕРЖКИ ДЛЯ ЛЮДЕЙ С ПСИХ...  2021-08-18   
9541                                          АНО «КЦМ»  2021-08-24   
9542                                      АНО "ГОРНЭКС"  2021-08-24   
9543  АНО "ЦЕНТР ЭКОЛОГИЧЕСКИХ И СОЦИАЛЬНЫХ ИНИЦИАТИ...  2021-08-25   
9544             БЛАГОТВОРИТЕЛЬНЫЙ ФОНД "ЗДРАВИЯ ЖЕЛАЮ"  2021-09-08   

                                              mainOkved  \
9540  Предоставление прочих социальных услуг без обе...   
9541  Деятельность по организации конференций и выст...   
9542                   Судебно-экспертная деятельность}   
9543                            Сбор неопасных отходов}   
9544  Предоставление прочих финансовых услуг, кроме ...   

                                opfCombined  \
9540  Автономные некоммерческие организации   
9541  Автономные некоммерческие организации   
9542  Автономные некоммерческие организации   
9543  Автономные некоммерческие организации   
9544                Благотворительные фонды   

                                          okvedCombined  vedSimple  \
9540              Автономные некоммерческие организации        АНО   
9541  Деятельность по организации конференций и выст...        NaN   
9542                    Судебно-экспертная деятельность        NaN   
9543                             Сбор неопасных отходов        NaN   
9544  Предоставление прочих финансовых услуг, кроме ...    Финансы   

      websiteCreationDate  website_country      website_hostname  \
9540  2021-08-17 15:54:57   United Kingdom                  None   
9541   2021-09-28 7:05:18           Russia         vh290.sweb.ru   
9542  2021-09-02 13:05:33           Russia                  None   
9543  2021-09-27 11:38:03           Russia        parking.reg.ru   
9544  2021-09-20 11:42:29           Russia  wcarp.hosting.nic.ru   

          website_isp  
9540             None  
9541             None  
9542   Ddos-guard LTD  
9543             None  
9544  JSC "RU-CENTER"  

[5 rows x 30 columns]

How many Russian NGOs host their websites on foreign servers right now as of 28.06.2022?

In [4]:
#concat_ngos[concat_ngos['website_country'] != 'Russia' and concat_ngos['website_country'] != 'Russian Federation'].count()
len(concat_ngos[(concat_ngos['website_country'] != 'Russia') & (concat_ngos['website_country'] != 'Russian Federation')  & ~concat_ngos['website_country'].isnull()])

998

**All in all 933 NGOs as of 28.06.2022 host websites on non-Russian hostings**

In [ ]:
unavailable_sites

['vsevfootball.ru',
 'ppktdk.ru',
 'kso.org.ru',
 'держимсявместе.рф',
 'ilma095.ru',
 'paleoethnology.ru',
 'ekspert55.ru',
 'политехник27.рф',
 'атмосфера-жизни.рф',
 'конкурсы-твояприрода.рф',
 'svetserdec.ru',
 'china-shina.ru',
 'teplonashihryk.ru',
 'iledu.ru',
 'arcticcouncil.ru',
 'арктический-совет.рф',
 'арктическийсовет.рф',
 'rus-prazd.ru',
 'мы-спб.рф',
 'patriot-wtf.ru',
 'ano-mcdpo.ru',
 'sevfvs.ru',
 'siberian-patrio.ru',
 'fond-udprf.ru',
 'pochinokfund.ru',
 'fsk-junior.ru',
 'dbfa.ru',
 '3d-paint.ru',
 'анолады.рф',
 'mooiniciativa.ru',
 'аномокит.рф',
 'транспортная-безопасность.рф',
 'canistherapya.ru',
 'fondboks12.ru',
 'nkoparadigma.ru',
 'koap25.ru',
 'soczabota67.ru',
 'penza-hor.ru',
 'kobzonfond.ru',
 'росбол.рф',
 'роои-созидание.рф',
 'anociup.ru',
 'bizline18.ru',
 'www.crot-edu.ru',
 'prostymi.ru',
 'lifeline-spb.ru',
 'anoalliance.ru',
 'anolliance.ru',
 'сроосфэг.рф',
 'history-ng.ru',
 'фондвозрождения.рф',
 'mertun.ru',
 'fci-sreda.ru',
 'светлая-гор

In [ ]:
len(unavailable_sites)

146

146 websites have invalid links

### Save data

In [ ]:
concat_ngos.to_csv("ngos_hostings.csv", encoding='utf-8')

### Second run just in case

In [3]:
concat_ngos = pd.read_csv("ngos_hostings_countries.csv", encoding='utf-8')
concat_ngos.head()

Unnamed: 0  Unnamed: 0.1           ogrn         website  bin_robots  \
0           0             0  1026800000666    profobr68.ru         1.0   
1           1             1  1022900002861  проф-севмаш.рф         1.0   
2           2             2  1025200008140      vdpor52.ru         1.0   
3           3             3  1037739471220     orthonet.ru         0.0   
4           4             4  1085200004450   vdpo-sarov.ru         1.0   

   bin_sitemap  bin_fb  bin_vk  bin_ig  bin_ok  ...  \
0          1.0     0.0     0.0     0.0     0.0  ...   
1          0.0     0.0     1.0     0.0     0.0  ...   
2          1.0     0.0     0.0     0.0     0.0  ...   
3          0.0     0.0     0.0     0.0     0.0  ...   
4          1.0     0.0     1.0     0.0     0.0  ...   

                                           shortName     regDate  \
0  ТАМБОВСКАЯ ОБЛАСТНАЯ ОРГАНИЗАЦИЯ ОБЩЕРОССИЙСКО...  1938-11-10   
1                  МОО - ППО АО "ПО "СЕВМАШ" СУДПРОФ  1942-03-15   
2  НИЖЕГОРОДСКОЕ ОБЛАСТНОЕ ОТДЕЛЕНИЕ ВДПО, ВДПО Н...  1958-06-20   
3                              СОЮЗ ПИСАТЕЛЕЙ РОССИИ  1958-12-07   
4                                   САРОВСКОЕ "ВДПО"  1962-03-30   

                                           mainOkved  \
0               Деятельность профессиональных союзов   
1               Деятельность профессиональных союзов   
2  Деятельность прочих общественных организаций, ...   
3  Деятельность профессиональных членских организ...   
4  Деятельность прочих общественных организаций, ...   

                opfCombined  \
0   Профсоюзные организации   
1   Профсоюзные организации   
2  Общественные организации   
3  Общественные организации   
4  Общественные организации   

                                       okvedCombined  \
0               Деятельность профессиональных союзов   
1               Деятельность профессиональных союзов   
2  Деятельность прочих общественных организаций, ...   
3  Деятельность профессиональных членских организ...   
4  Деятельность прочих общественных организаций, ...   

                  vedSimple  websiteCreationDate  website_country  \
0                 Профсоюзы  2016-01-15 19:31:34           Russia   
1                 Профсоюзы  2018-08-27 11:35:06           Russia   
2  Общественные организации  2016-09-28 12:05:06           Russia   
3                 Профсоюзы  2005-11-24 21:00:00           Russia   
4  Общественные организации   2021-04-15 8:10:11           Russia   

     website_hostname  website_isp  
0  m2.wolf3.beget.com          NaN  
1    vh134.timeweb.ru          NaN  
2            websg.ru          NaN  
3  unspecified.mtw.ru          NaN  
4                 NaN          NaN  

[5 rows x 32 columns]

In [ ]:
import socket
from socket import gaierror


for i, entry in enumerate(concat_ngos.itertuples()):
  if isinstance(entry.website_country, float):
    # Get IP address
    try:
      ip_address = socket.gethostbyname(entry.website)
    except gaierror:
      continue

    print(i, entry.website)
    country_handler = Country(ip_address)
    country = country_handler.get_country(hostname=False)
    print(country)

    concat_ngos.at[i, "website_country"] = country

336 www.mofk.net
United States
518 kimvestnik.ru
Russian Federation
521 tofpmp.ru
Russian Federation
524 intellectdm.ru
Russian Federation
528 мпи.рф
Russian Federation
682 no-vpered.ru
Russian Federation
685 bel-pobeda.ru
Russian Federation
686 znamya31.ru
United States
689 kurortcenter.ru
Russian Federation
696 vesti-m.ru
Russian Federation
698 empatiann.ru
Russian Federation
699 vdpo33.ru
Russian Federation
880 www.sefer.ru
Russian Federation
881 s-kub.ru
Russian Federation
884 astrakhancy.ru
Russia
1672 russkiebezrossii.ru
Russian Federation
2349 www.infars.ru
Russian Federation
2353 www.unitalm.ru
Russian Federation
2356 guic.ru
Russian Federation
2623 45led.ru
Russian Federation
2707 wogart.narod.ru
Virgin Islands, British
2708 anoavtoshkola.ru
United States
2712 basmania.ru
Russia
2716 vdpo-ku.com
Israel
2720 istokmw.ru
Russian Federation
2722 cnsio.ru
United Kingdom
2723 dlokean-khv.com
Israel
3032 bushido.ru
United States
3035 loovoi.usite.pro
Russian Federation
3804 academpar

In [ ]:
concat_ngos[concat_ngos["website"] == "www.mofk.net"]

Unnamed: 0           ogrn       website  bin_robots  bin_sitemap  bin_fb  \
336         336  1027739651873  www.mofk.net         1.0          1.0     0.0   

     bin_vk  bin_ig  bin_ok  bin_youtube  ...  shortName     regDate  \
336     0.0     0.0     0.0          0.0  ...        NaN  1992-02-10   

      mainOkved         opfCombined  \
336  None, code  Общественные фонды   

                                         okvedCombined  vedSimple  \
336  Деятельность профессиональных членских организ...  Профсоюзы   

     websiteCreationDate  website_country  website_hostname website_isp  
336  2017-01-03 19:33:54    United States               NaN         NaN  

[1 rows x 31 columns]

In [ ]:
concat_ngos.to_csv("ngos_hostings_2.csv", encoding='utf-8')

### testing (not nessesary to run)

In [1]:
import ipinfo
from util import access_token_ipinfo


handler = ipinfo.getHandler(access_token_ipinfo)

ModuleNotFoundError: ignored

In [ ]:
#ip_address =
details = handler.getDetails("194.58.90.161")
details.all

{'city': 'Moscow',
 'country': 'RU',
 'country_name': 'Russia',
 'hostname': 'websg.ru',
 'ip': '194.58.90.161',
 'latitude': '55.7522',
 'loc': '55.7522,37.6156',
 'longitude': '37.6156',
 'org': 'AS197695 Domain names registrar REG.RU, Ltd',
 'postal': '101000',
 'region': 'Moscow',
 'timezone': 'Europe/Moscow'}

Question: hostname is sometimes different from registrar/org.

Answer: we need both

In [ ]:
from pprint import pprint

# test batch hosting check
pprint(handler.getBatchDetails(['1.1.1.1','8.8.8.8','1.2.3.4/country']))

{'1.1.1.1': {'anycast': True,
             'city': 'Los Angeles',
             'country': 'US',
             'country_name': 'United States',
             'hostname': 'one.one.one.one',
             'ip': '1.1.1.1',
             'latitude': '34.0522',
             'loc': '34.0522,-118.2437',
             'longitude': '-118.2437',
             'org': 'AS13335 Cloudflare, Inc.',
             'postal': '90076',
             'region': 'California',
             'timezone': 'America/Los_Angeles'},
 '1.2.3.4/country': 'AU',
 '8.8.8.8': {'anycast': True,
             'city': 'Mountain View',
             'country': 'US',
             'country_name': 'United States',
             'hostname': 'dns.google',
             'ip': '8.8.8.8',
             'latitude': '37.4056',
             'loc': '37.4056,-122.0775',
             'longitude': '-122.0775',
             'org': 'AS15169 Google LLC',
             'postal': '94043',
             'region': 'California',
             'timezone': 'America/Lo